### ISSUE TITLE
docs: Issue in page Nodetool refresh

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#13041]
I would like to report an issue in page https://docs.scylladb.com/branch-5.1/operating-scylla/nodetool-commands/refresh

### Problem

It doesn't mention that changing gc_grace_seconds prior to uploading sstables is critical for ensuring data consistency. Otherwise, some of the tombstones may be garbage-collected without being compacted with data they cover which may arrive later. 

###  Suggest a fix

Mention that gc_grace_seconds should be increased prior to the upload to a value of 31557600 to prevent tombstone gc during upload. After the upload, it should be restored to the normal value.

You don't have to do it for tables which are created with `tombstone_gc = {'mode':'repair'};`, unless the tables were repaired in the cluster prior to the upload.

**Comment by fee-mendes** [comment]
And please add a red visible warning :-D

**Comment by Michal-Leszczynski** [comment]
What about loading SSTables from let's say `system_schema.tables`?
When I try to run an alter query like:  `ALTER TABLE system_schema.tables WITH gc_grace_seconds = {gc_grace_seconds value}`, 
I get the error: `Unauthorized: Error from server: code=2100 [Unauthorized] message="system_schema keyspace is not user-modifiable."`

**Comment by tgrabiec** [comment]
@Michal-Leszczynski You alter only tables for which you are uploading sstables. You should not use `nodetool refresh` on system tables, so not on system_schema.

**Comment by tgrabiec** [comment]
@bhalevy If we go this direction, I think it makes sense if the manager or any outside user knows as little as possible about what is necessary. So we should have a general `before_upload` and `after_upload`, and scylla does what is necessary, like disabling tombstone gc. And upload can fail if it's not within these.

**Comment by Michal-Leszczynski** [comment]
I use load and stream (storage service API) for system tables when restoring backup in Scylla Manager universal restore, but in case of system_schema.* tables I don't alter their ggs value, because of the error mentioned above.

I see that all system_schema.* tables have ggs set to `604800` and since schema SSTables should be rather small (meaning that restoring them shouldn't take much time), is it safe to use load and stream without modifying the ggs value?

**Comment by tgrabiec** [comment]
Do you restart nodes after that? Schema won't take effect if you just stream the sstables. Also, a schema change made after you stream those sstables but before you restart the nodes can put the node in an inconsistent state because the in-memory schema will not reflect the on-disk state. In other nodes, schema is not supposed to be changed by updating local system tables with the node up, but through the CQL path, or with the node being down.  

604800 is 10 days. It's not safe, because the sstables may contain tombstones which are older than 10 days. Even if sstables are small, there is no guarantee that they all will be compacted together.

**Comment by Michal-Leszczynski** [comment]
Yes, I make a rolling restart of an entire cluster.
In case of SM restore procedure, it can be assumed that the cluster (that we restore the schema into) does not have any overlapping schema, it is basically a fresh cluster. If compaction is the problem, would disabling autocompaction (column family API) solve this issue? I see that it can be performed on system_schema.* tables.

**Comment by tgrabiec** [comment]
Disabling compaction may not cover all the cases, e.g. if you expire writes when flushing memtables.

In general, the manager should not try to achieve consistency by relying on implementation details and disable individual processes in scylla. Even if this would work now, it can easily break in the future once the internals change. Scylla should expose an API which would specifically signal this intent, and take care of upholding to the promise in whatever way is necessary.

**Comment by Michal-Leszczynski** [comment]
Thanks for the explanation!
And just to clarify, up to this point, Scylla Manager was using the ansible restore script for restoring both schema (SM backups and restores schema in form of SSTables) and data into a fresh cluster. The script can be found [here](https://github.com/scylladb/scylla-manager/blob/master/ansible/restore/restore.yaml). In this case, there is a requirement that the backed-up cluster and the cluster that we are restoring the data into have the same number of nodes. This allows us to map each backed-up node to one node in destination cluster.
The procedure looks like that:
- stop Scylla services
- wipe all cluster data and restore the original token ring
- start Scylla services
- download both schema and data into upload dirs (given node gets SSTables from the backed-up node that it is mapped to and according to restored token ring)
- use nodetool refresh on system_schema tables (no load and stream)
- restart cluster
- use nodetool refresh on remaining data tables (no load and stream)

I don't see any code that would alter gc grace seconds or compaction in this procedure. So my question is, is this ansible script safe, or can it also fail because of the issues that you mentioned above?

**Comment by tgrabiec** [comment]
There is one potential issue with this, and it's a bit different than what was described earlier here. Depending on how you choose sstables a given new replica receives. Is it only a subset of sstables of the old cluster which belongs to the old replica, or are theses all sstables? 

If it's only a subset, and the snapshot is not in a fully repaired state (it's hard to guarantee that this is the case), then it may happen that only one replica of some key has a deletion which is now expired, but was not replicated to other replicas. After you attach the sstables, that tombstone may get garbage-collected before it's replicated to other replicas, and cause resurrection of deleted data which is still present on those other replicas which missed the deletion.

So in addition to changing gc_grace_seconds, the operator should repair the cluster before gc_grace_seconds is restored to the normal value.

**Comment by Michal-Leszczynski** [comment]
This procedure restores all SSTables.
But still, as you said, despite restoring all SSTables (which only influences the lack of need to repair), this procedure requires the addition of altering gc grace seconds and without it, the data resurrection problem may occur?

**Comment by tgrabiec** [comment]
On Sun, Mar 5, 2023 at 6:09 PM Michal-Leszczynski ***@***.***>
wrote:

> This procedure restores all SSTables.
>

But does each new replica receive the same copy of sstables, from all of
the old replicas?

**Comment by Michal-Leszczynski** [comment]
Let's say that the backed-up cluster consisted of 3 hosts A1, B1, C1 and that we want to restore it into a new cluster A2, B2, C2. We map A1 - A2, B1 - B2, C1 - C2, so the A2 will receive all SSTables backed-up from the A1 node, but won't get any SSTables from B1 or C1, those will go only to B2 and C2 regardless of the RF. Does this answer your question?

**Comment by tgrabiec** [comment]
> Let's say that the backed-up cluster consisted of 3 hosts A1, B1, C1 and that we want to restore it into a new cluster A2, B2, C2. We map A1 - A2, B1 - B2, C1 - C2, so the A2 will receive all SSTables backed-up from the A1 node, but won't get any SSTables from B1 or C1, those will go only to B2 and C2 regardless of the RF. Does this answer your question?

Yes, that's what I thought. In this case lack of repair is problematic (we should both adjust gc grace seconds and repair before restoring it to the previous value). A1 and B1 may contain an expired deletion which is not in C1. If you don't repair, A2 and B2 may compact away the deletion before the next scheduled repair and deleted data which is still in C2 may get resurrected.

**Comment by bhalevy** [comment]
I agree with Tomek. The reasoning behind gc_grace_seconds is to give the admin time to run repair in that period.  It doesn't guarantee anything else.
With `repair` garbage collection mode, things are different. Tombstones won't be purged until a later repair successfully repaired the respective token.
We compare the tombstone `deleted_at` time point against the repair time to determine if repair ran after the tombstone was written or before it.
@asias, in load_and_stream mode, do we "refresh" tombstones deletion time to the current time, or do we restore the original time?
What about ttl expiration times?
When data expires it effectively becomes a tombstone and the expiration time becomes the deletion time.
How do we treat that in load_and_stream?

**Comment by asias** [comment]
> I would like to report an issue in page https://docs.scylladb.com/branch-5.1/operating-scylla/nodetool-commands/refresh
> ### Problem
> 
> It doesn't mention that changing gc_grace_seconds prior to uploading sstables is critical for ensuring data consistency. Otherwise, some of the tombstones may be garbage-collected without being compacted with data they cover which may arrive later and lead to data resurrection.
> ### Suggest a fix
> 
> Mention that gc_grace_seconds should be increased prior to the upload to a value of 2147483647 to prevent tombstone gc during upload. After the upload, it should be restored to the normal value.
> 
> When not using load-and-stream, the operator should also run repair before restoring the gc grace seconds to the normal value.
> 
> You don't have to do it for tables which are created with `tombstone_gc = {'mode':'repair'};`, unless the tables were repaired in the new cluster prior to the upload (when uploading into a non-fresh cluster).

Instead of setting gc_grace_seconds to a big number 2**31 -1, we can set 

tombstone_gc = {'mode':'disabled'}

to disable  compaction during loading of data into the cluster and set back to

tombstone_gc = {'mode':'timeout'}

when loading of data is done.

Even with tombstone_gc = {'mode':'repair'}, as you mentioned above we need to set tombstone_gc = {'mode':'disabled'} to if there was repair already performed in the cluster.

So, to both cases, it is simpler and safer to just ask user to set tombstone_gc = {'mode':'disabled'} during loading of data and revert back to either timeout mode or repair mode.

**Comment by asias** [comment]
> I agree with Tomek. The reasoning behind gc_grace_seconds is to give the admin time to run repair in that period. It doesn't guarantee anything else. With `repair` garbage collection mode, things are different. Tombstones won't be purged until a later repair successfully repaired the respective token. We compare the tombstone `deleted_at` time point against the repair time to determine if repair ran after the tombstone was written or before it. @asias, in load_and_stream mode, do we "refresh" tombstones deletion time to the current time, or do we restore the original time? What about ttl expiration times? When data expires it effectively becomes a tombstone and the expiration time becomes the deletion time. How do we treat that in load_and_stream?

No, load and stream does not rewrite the deletion time. It writes what is read without any modification.

**Comment by Michal-Leszczynski** [comment]
I see that altering `tombstone_gc = {'mode':'disabled'}` also does not work on system_schema.* tables.
So is there any way in which I can safely restore backed-up schema from SSTables?
This is actually problematic on SM side, because:
- SM cannot trust the output of `DESC SCHEMA`
> 	// Some schema changes, like dropping columns, are applied lazily to
> 	// sstables during compaction. Information about those schema changes is
> 	// recorded in the system schema tables, but not in the output of "desc schema".
> 	// Using output of "desc schema" is not enough to restore all schema changes.
> 	// As a result, writes in sstables may be incorrectly interpreted.
> 	// For example, writes of deleted columns which were later recreated may be
> 	// resurrected.
- SM cannot alter `ggs` or `tombstone_gc`, so using load and stream is not safe

So is there any way in which I can safely restore schema?

**Comment by tgrabiec** [comment]
If there was no schema disagreement at the time of taking the backup then we're fine, no need for repair or gc grace seconds tricks on schema tables. Unlike user tables, schema tables automatically repair themselves across nodes. 

If backup was taken during schema change and there is disagreement between nodes, we may run into the problems mentioned earlier. You could avoid this by giving each node sstables from all the nodes when restoring schema, simulating repair.

**Comment by Michal-Leszczynski** [comment]
Just to clarify, how can I perform "giving each node sstables from all the nodes" in a safe way?
Do you mean a procedure like this one:
* For each node D in restore destination cluster:
   * For each node S from backed-up cluster: 
      * For each backed-up schema table T from node S:
         * download schema SSTables from T `into node D` upload dir
         * use `nodetool refresh`

Because SSTables from different backed-up nodes might have overlapping names, I can't use nodetool refresh on all SSTables in one call - I have to divide that between backed-up nodes.

Can this be simplified to the following procedure:
* For each node S from backed-up cluster: 
   * For each backed-up schema table T from node S:
      * download schema SSTables from T `into any node in destination cluster` upload dir
      * use  load and stream` with `primary_replica_only` set to false

**Comment by tgrabiec** [comment]
If you call `nodetool refresh` multiple times on a given replica, atomic upload is not guaranteed, and we would need to disable tombstone gc around all refreshes to guarantee safety.

Using `load and stream` on local system tables won't replicate the data to other nodes, if I read the code correctly, since the tables use local replication strategy, not everywhere replication strategy.

**Comment by bhalevy** [comment]
> You could play with renaming sstable files to avoid name conflicts.

You mean meddling with the sstable generations, right?

**Comment by fee-mendes** [comment]
> If there was no schema disagreement at the time of taking the backup then we're fine, no need for repair or gc grace seconds tricks on schema tables. Unlike user tables, schema tables automatically repair themselves across nodes.

And on top of that: @Michal-Leszczynski note that as you are doing a full cluster restore, the process to restore `system_schema` assumes the replicas are down. So you restore all source `system_schema`'s to their respective replicas, bring the nodes up, and this should be already enough to get the cluster into agreement state when all replicas come up - one at a time - and start negotiating their schemas.

**Comment by tgrabiec** [comment]
On Fri, Mar 10, 2023 at 2:57 PM Benny Halevy ***@***.***>
wrote:

> You could play with renaming sstable files to avoid name conflicts.
>
> You mean meddling with the sstable generations, right?
>

Yes.
